src/Client.cpp

```diff
@@ -492,32 +493,53 @@ ::initBeforeLoop()
 
 			if (g_fds_array[ifd]->sock_type == SOCK_STREAM) {
 				log_dbg ("[fd=%d] Connecting to: %s:%d...", ifd, inet_ntoa(g_fds_array[ifd]->server_addr.sin_addr), ntohs(g_fds_array[ifd]->server_addr.sin_port));
+
+
 				if (connect(ifd, (struct sockaddr*)&(g_fds_array[ifd]->server_addr), sizeof(struct sockaddr)) < 0) {
 					if (os_err_in_progress()) {
-						fd_set rfds, wfds;
-						struct timeval tv;
-						tv.tv_sec = 0; tv.tv_usec = 500;
-
-						FD_ZERO(&rfds);
-						FD_ZERO(&wfds);
-
-						int max_fd = -1;
-
-						FD_SET(ifd, &wfds);
-						FD_SET(ifd, &rfds);
-						if(ifd > max_fd) max_fd = ifd;
-
-						select(max_fd + 1, &rfds, &wfds, NULL, &tv);
-						if(FD_ISSET(ifd, &wfds) || FD_ISSET(ifd, &rfds)) {
-							socklen_t err_len;
-							int error;
-
-							err_len = sizeof(error);
-							if (getsockopt(ifd, SOL_SOCKET, SO_ERROR, &error, &err_len) < 0 || error != 0) {
-								log_err("Can`t connect socket");
-								rc = SOCKPERF_ERR_SOCKET;
-								break;
-						    }
+#ifdef  USING_VMA_EXTRA_API
+						if (g_pApp->m_const_params.is_vmapoll && g_vma_api) {
+							int ring_fd = 0;
+							int poll = 0;
+							g_vma_api->get_socket_rings_fds(ifd, &ring_fd,1);
+							assert((-1) != ring_fd);
+							while (poll == 0) {
+								struct vma_completion_t vma_comps;
+								poll = g_vma_api->vma_poll(ring_fd, &vma_comps, 1, 0);
+								if (poll > 0) {
+									if (vma_comps.events & EPOLLOUT) {
+										ifd = vma_comps.user_data;
+									}
+								}
+							}
+						}else
```

Comment:
I agree with @avnerbh. I think we could implement static functions as _connect_check() and _connect_check_vma() and call
 ```
#ifdef  USING_VMA_EXTRA_API
   _connect_check_vma();
#else
   _connect_check();
#endif
```